python/ray/data/_internal/execution/operators/map_data_processor.py

```diff
@@ -0,0 +1,317 @@
+    ) -> Iterable[Block]:
+        iter = input_blocks
+        # Apply the transform functions sequentially to the input iterable.
+        for transform_fn in self._transform_fns:
+            iter = transform_fn(iter, ctx)
+        return iter
+
+    def fuse(self, other: "MapDataProcessor") -> "MapDataProcessor":
+        """Fuse two MapDataProcessors together."""
+
+        def fused_init_fn():
+            self._init_fn()
+            other._init_fn()
+
+        # TODO(Scott): Add zero-copy batching between transform functions.
+        fused_transform_fns = self._transform_fns + other._transform_fns
+        return MapDataProcessor(fused_transform_fns, init_fn=fused_init_fn)
+
+
+# Below are util functions for converting input/output data.
+
+
+def _input_blocks_to_rows_fn(blocks: Iterable[Block], _: TaskContext) -> Iterable[Row]:
+    """Converts input blocks to rows."""
+    for block in blocks:
+        block = BlockAccessor.for_block(block)
+        for row in block.iter_rows(public_row_format=True):
+            yield row
+
+
+_input_blocks_to_rows = MapTransformFn(
+    _input_blocks_to_rows_fn,
+    MapTransformFnDataType.Block,
+    MapTransformFnDataType.Row,
+)
+
+
+def _input_blocks_to_batches(
+    blocks: Iterable[Block],
+    _: TaskContext,
+    batch_size: Optional[int] = None,
+    batch_format: str = "default",
+    zero_copy_batch: bool = False,
+) -> Iterable[DataBatch]:
+    """Converts input blocks to batches."""
+    block_iter = iter(blocks)
+    first = next(block_iter, None)
+    if first is None:
+        return []
+    blocks = itertools.chain([first], block_iter)
+    empty_block = BlockAccessor.for_block(first).slice(0, 0, copy=True)
+    # Don't hold the first block in memory, so we reset the reference.
+    first = None
+
+    # Ensure that zero-copy batch views are copied so mutating UDFs don't error.
+    ensure_copy = not zero_copy_batch and batch_size is not None
+    formatted_batch_iter = batch_blocks(
+        blocks=blocks,
+        stats=None,
+        batch_size=batch_size,
+        batch_format=batch_format,
+        ensure_copy=ensure_copy,
+    )
+
+    first = next(formatted_batch_iter, None)
+    if first is None:
+        # If the input blocks are all empty, then yield an empty block with same
+        # format as the input blocks.
+        return [empty_block]
+    else:
+        return itertools.chain([first], formatted_batch_iter)
+
+
+def _to_output_blocks(
+    iter, _: TaskContext, iter_type: MapTransformFnDataType
+) -> Iterable[Block]:
+    """Convert UDF-returned data to output blocks."""
+    output_buffer = BlockOutputBuffer(
+        None, DataContext.get_current().target_max_block_size
+    )
+    for data in iter:
+        if iter_type == MapTransformFnDataType.Block:
+            output_buffer.add_block(data)
+        elif iter_type == MapTransformFnDataType.Batch:
+            output_buffer.add_batch(data)
+        else:
```

Comment:
nit: move these checks to outside the for loop so you only need to do it once. The assumption for this method is that we're not mixing and matching Block, Row, and DataBatch right?